use latest ouroboros-network

avieth opened the pull request:
Main points:

- `cardano-ledger` was required to update, and it no longer supports the `GCSpec` configuration. No problem: exit if the cardano-sl configuration uses one, telling the user they must use a json genesis file.
- This `ResourceRegistry` is now ubiquitous, and it's tossed me into the pit of failure. I made *one* and used it to make a `ChainDB`, as well as use that `ChainDB`, ~but now the executable dies with `ResourceRegistryUsedFromDifferentThread`~ that was the index not forking a linked one. I fixed that, but uses of `streamBlocks` and `newBlockReader` in the `cardano-sl` logic layer will certainly fail with the same error. Not sure how to proceed.

avieth committed:
use latest ouroboros-network

big fallout

avieth committed:
properly use resource registry for index

on trackChainDB

It forced me to re-write it to not take a continuation, because the
thread that creates the reader _must_ be spawned by forkLinkedTransfer

mrBliss commented on src/exec/DB.hs:
```diff
@@ -127,5 +127,6 @@ withDB dbOptions dbTracer indexTracer rr securityParam nodeConfig extLedgerState
-    Sqlite.withIndexAuto epochSlots indexTracer (indexFilePath dbOptions) $ \idx ->
-      Index.trackChainDB rr idx cdb $ k idx cdb
+    Sqlite.withIndexAuto epochSlots indexTracer (indexFilePath dbOptions) $ \idx -> do
+      _ <- forkLinkedTransfer rr $ \rr' -> Index.trackChainDB rr' idx cdb
```
Instead of `forkLinkedTransfer`, you can create a new resource registry for `Index.trackChainDB`as you will only use it for simple resources, not for threads. And as this is a scoped context you can use `with` and don't have to link to the parent registry.

So:
```haskell
ResourceRegistry.with $ \rr' -> Index.trackChainDB rr' idx cdb
```

mrBliss replied:
Of course, you had no way of knowing this, so I'll document this.

mrBliss committed:
Remove redundant import

mrBliss committed:
Correct ResourceRegistry usage

The resources don't have to be transferred to the top-level registry when the
thread terminates, so `forkLinked` is enough. Create a fresh registry for the
lifetime of the thread, as the allocated resources should be released when the
thread terminates.

mrBliss commented:
I'll update this PR.

I have noticed that `forkLinked(Transfer)` requires 2x <kbd>CTRL-C</kbd> to shut down, so there is something going wrong. I'll debug.

mrBliss commented:
> I'll update this PR.
> 
> I have noticed that `forkLinked(Transfer)` requires 2x CTRL-C to shut down, so there is something going wrong. I'll debug.

Since Edsko's working on fixing bugs in the `ResourceRegistry` (the upcoming https://github.com/input-output-hk/ouroboros-network/pull/929/ is just the start), I'll leave it as is. When Edsko's work is merged, I'll revisit this.

However, in the meantime, this PR can be merged. Anybody blocked on it is free to bors it.

nc6 commented:
bors r+
